client/batch.go

```diff
@@ -0,0 +1,426 @@
+// encoding.BinaryMarshaler. value can be any key type or a proto.Message.
+func (b *Batch) CPut(key, value, expValue interface{}) *Batch {
+	b = b.ensureNotNil()
+
+	k, err := marshalKey(key)
+	if err != nil {
+		b.initResult(0, 1, err)
+		return b
+	}
+	v, err := marshalValue(value)
+	if err != nil {
+		b.initResult(0, 1, err)
+		return b
+	}
+	ev, err := marshalValue(expValue)
+	if err != nil {
+		b.initResult(0, 1, err)
+		return b
+	}
+	b.calls = append(b.calls, ConditionalPut(proto.Key(k), v, ev))
+	b.initResult(1, 1, nil)
+	return b
+}
+
+// Inc increments the integer value at key. If the key does not exist it will
+// be created with an initial value of 0 which will then be incremented. If the
+// key exists but was set using Put or CPut an error will be returned.
+//
+// A new result will be appended to the batch which will contain a single row
+// and Result.Err will indicate success or failure.
+//
+// key can be either a byte slice, a string, a fmt.Stringer or an
+// encoding.BinaryMarshaler.
+func (b *Batch) Inc(key interface{}, value int64) *Batch {
+	b = b.ensureNotNil()
+
+	k, err := marshalKey(key)
+	if err != nil {
+		b.initResult(0, 1, err)
+		return b
+	}
+	b.calls = append(b.calls, Increment(proto.Key(k), value))
+	b.initResult(1, 1, nil)
+	return b
+}
+
+// Scan retrieves the rows between begin (inclusive) and end (exclusive).
+//
+// A new result will be appended to the batch which will contain up to maxRows
+// rows and Result.Err will indicate success or failure.
+//
+// key can be either a byte slice, a string, a fmt.Stringer or an
+// encoding.BinaryMarshaler.
+func (b *Batch) Scan(s, e interface{}, maxRows int64) *Batch {
+	begin, err := marshalKey(s)
+	if err != nil {
+		b.initResult(0, 0, err)
+		return b
+	}
+	end, err := marshalKey(e)
+	if err != nil {
+		b.initResult(0, 0, err)
+		return b
+	}
+	b.calls = append(b.calls, Scan(proto.Key(begin), proto.Key(end), maxRows))
+	b.initResult(1, 0, nil)
+	return b
+}
+
+// Del deletes one or more keys.
+//
+// A new result will be appended to the batch and each key will have a
+// corresponding row in the returned Result.
+//
+// key can be either a byte slice, a string, a fmt.Stringer or an
+// encoding.BinaryMarshaler.
+func (b *Batch) Del(keys ...interface{}) *Batch {
+	b = b.ensureNotNil()
+
+	var calls []Call
+	for _, key := range keys {
+		k, err := marshalKey(key)
+		if err != nil {
+			b.initResult(0, len(keys), err)
+			return b
+		}
+		calls = append(calls, Delete(proto.Key(k)))
+	}
+	b.calls = append(b.calls, calls...)
+	b.initResult(len(calls), len(calls), nil)
+	return b
+}
+
+// DelRange deletes the rows between begin (inclusive) and end (exclusive).
+//
+// A new result will be appended to the batch which will contain 0 rows and
+// Result.Err will indicate success or failure.
+//
+// key can be either a byte slice, a string, a fmt.Stringer or an
+// encoding.BinaryMarshaler.
+func (b *Batch) DelRange(s, e interface{}) *Batch {
```

Comment:
A client which touches a fixed set of keys could use chaining, but it isn't
necessarily clearer:

err := db.Run(client.B().Put("a", "1").Put("b", "2"))


vs

b := &client.Batch{}
b.Put("a", "1")
b.Put("b", "2")
err := db.Run(b)


On Tue, Jun 9, 2015 at 6:19 PM, Tobias Schottdorf <notifications@github.com>
wrote:

> In client/batch.go
> <https://github.com/cockroachdb/cockroach/pull/1318#discussion_r32071102>:
>
> > +		}
> > +		calls = append(calls, Delete(proto.Key(k)))
> > +	}
> > +	b.calls = append(b.calls, calls...)
> > +	b.initResult(len(calls), len(calls), nil)
> > +	return b
> > +}
> > +
> > +// DelRange deletes the rows between begin (inclusive) and end (exclusive).
> > +//
> > +// A new result will be appended to the batch which will contain 0 rows and
> > +// Result.Err will indicate success or failure.
> > +//
> > +// key can be either a byte slice, a string, a fmt.Stringer or an
> > +// encoding.BinaryMarshaler.
> > +func (b *Batch) DelRange(s, e interface{}) *Batch {
>
> Wouldn't any client code that touches a fixed set of keys use the chaining?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1318/files#r32071102>.
>